pynav/lib/models.py

```diff
@@ -82,6 +81,165 @@ def covariance(
+        """Computes the discrete-time Jacobian.
+
+        Here, we start in continuous-time and discritize the linearized
+        continuous-time system.
+
+        Parameters
+        ----------
+        x : IMUState
+            The state at which to evaluate the Jacobian at.
+        u : StampedValue
+            Measurement at which to evaluate the Jacobian at.
+        dt : float
+            Discritization timestep.
+
+        Returns
+        -------
+        np.ndarray
+            Process model Jacobian
+        """
+        # Compute unbiased gyro and accel measurements using bias estimates
+
+        gyro_m = u.value[0:3, [0]]
+        acc_m = u.value[3:6, [0]]
+
+        unbiased_gyro = gyro_m - x.bg
+        unbiased_accel = acc_m - x.ba
+
+        # Right-invariant Jacobians
+        A_ct = np.zeros((x.dof, x.dof))
+
+        # Extract relevant states
+        C = x.attitude
+        v = x.velocity
+        r = x.position
+        g_a = x.gravity
+
+        if x.direction == "left":
+            # TODO: These are the Jacobians for the right-invariant error,
+            # check that they are correct for a left perturbation
+            A_ct[0:3, 9:12] = -C
+            A_ct[3:6, 0:3] = SO3.cross(g_a)
+            A_ct[3:6, 9:12] = -SO3.cross(v) @ C
+            A_ct[3:6, 12:15] = -C
+            A_ct[6:9, 3:6] = np.identity(3)
+            A_ct[6:9, 9:12] = -SO3.cross(r) @ C
+        else:
+            raise NotImplementedError("Left jacobian not implemented.")
+
+        # Approximate matrix exponential to third-order
+        Adt = A_ct * dt
+        Adt_square = Adt @ Adt
+        Adt_cube = Adt_square @ Adt
+        Phi = np.identity(x.dof) + Adt + Adt_square / 2.0 + Adt_cube / 6.0
+
+        return Phi
+
+    def covariance(self, x: IMUState, u: StampedValue, dt: float) -> np.ndarray:
+        """Computes the discrete-time covariance.
+
+        Here again, we start in continuous-time and then discritize the
+        linear system.
+
+        Parameters
+        ----------
+        x : IMUState
+            _description_
+        u : StampedValue
+            _description_
+        dt : float
+            _description_
+
+        Returns
+        -------
+        np.ndarray
+            _description_
+        """
+
+        L_ct = np.zeros((x.dof, 12))
+
+        # Extract relevant states
+        C = x.attitude
+        v = x.velocity.reshape((-1, 1))
+        r = x.position.reshape((-1, 1))
+
+        if x.direction == "left":
+            L_ct[0:3, 0:3] = C
+            L_ct[3:6, 0:3] = SO3.cross(v) @ C
+            L_ct[3:6, 3:6] = C
+            L_ct[6:9, 0:3] = SO3.cross(r) @ C
+            L_ct[9:12, 6:9] = -np.identity(3)
+            L_ct[12:15, 9:12] = -np.identity(3)
+        else:
+            raise NotImplementedError("Left jacobian not implemented.")
+
+        # Compute discrete-time noise
+        return L_ct @ self._Q_c @ L_ct.T * dt
```

Comment:
i withdraw this comment. because you have to use `Q_c/dt` for noise generation you still have a quadratic relationship between noise-generation noise and process noise.